sklearn/naive_bayes.py

```diff
@@ -117,6 +118,313 @@ def predict_proba(self, X):
+
+        # Ensure class log priors are the same for all estimators
+        all_log_priors = np.hstack([all_log_priors])
+        if np.max(np.ptp(all_log_priors, axis=0)) < 1e-6:
+            log_prior = all_log_priors[0]
+        else:
+            raise ValueError("Class priors for every estimator "
+                             "must be the same.")
+
+        # Obtain the jll of each fitted estimator
+        jlls = [nb_model._joint_log_likelihood(
+                    np.array(_safe_indexing(X, cols, axis=1)))
+                for (_, nb_model, cols) in self.models_]
+
+        # Stack these jlls to give us
+        # the shape (estimator, sample, class)
+        jlls = np.hstack([jlls])
+
+        # Subtract the class log prior from all the jlls
+        # but add it back after the summation
+        jlls = jlls - log_prior
+        jll = jlls.sum(axis=0) + log_prior
+
+        return jll
+
+    @property
+    def _models(self):
+        return [(name, model) for name, model, _ in self.models]
+
+    @_models.setter
+    def _models(self, value):
+        self.models = [
+            (name, nb_model, cols) for ((name, nb_model), (_, _, cols))
+            in zip(value, self.models)]
+
+    def get_params(self, deep=True):
+        return self._get_params('_models', deep=deep)
+
+    def set_params(self, **kwargs):
+        self._set_params('_models', **kwargs)
+        return self
+
+    def _validate_models(self, X):
+
+        valid_modules = copy.copy(__all__)
+        valid_modules.remove("GeneralNB")
+        _dict_model = {}
+
+        _list_fit_prior = []
+        _list_class_prior = []
+
+        names, _, _ = zip(*self.models)
+        self._validate_names(names)
+
+        self._validate_column_callables(X)
+
+        # Check type
+        if not isinstance(self.models, list):
+            raise TypeError(
+                "Expected list but got {}".format(type(self.models)))
+
+        for model in self.models:
+
+            # Check type
+            if not isinstance(model, tuple):
+                raise TypeError(
+                    "Expected list of tuples "
+                    "but got list of {}s".format(type(model)))
+            if len(model) != 3:
+                raise ValueError("Expected tuple to have length of 3 "
+                                 "but got {}".format(len(model)))
+
+            name, estimator, cols = model
+
+            # Check naive bayes estimator
+            if callable(estimator):
+                raise ValueError("Wrong format specified.")
+            if not (hasattr(estimator, "fit")
+                    or hasattr(estimator, "_joint_log_likelihood")):
+                raise TypeError("Naive bayes estimator should implement "
+                                "the fit and _joint_log_likelihood methods. "
+                                "{} doesn't.".format(type(estimator)))
+            if estimator.__class__.__name__ not in valid_modules:
```

Comment:
Why do we need to enforce that the model is internal to scikit-learn?

Reply:
Only because we need the model to have implemented methods like `_joint_log_likelihood()` on top of the basic `fit()` and `predict()`. Should we relax this and just check for these attributes on the fly using `hasattr()`?

Reply:
Okay, removing.